TESTS/mbed_hal/lp_us_tickers/main.cpp

```diff
@@ -13,98 +13,416 @@
+
+    const uint32_t start = intf->read();
+
+    while (cycles -= step) {
+        /* Just wait. */
+    }
+
+    const uint32_t stop = intf->read();
+
+    core_util_critical_section_enter();
+
+    return (stop - start);
 }
 
-/* Ticker init should be run just once, thus read should always return a value that 
- * increases (no reset).
+void ticker_event_handler_stub(const ticker_data_t * const ticker)
+{
+    /* Indicate that ISR has been executed in interrupt context. */
+    if (IS_IRQ_MODE()) {
+        intFlag++;
+    }
+
+    /* Clear ticker interrupt. */
+    intf->clear_interrupt();
+}
+
+/* Test that ticker_init can be called multiple times and
+ * ticker_init resets the internal count and disables the ticker interrupt.
  */
-void test_ticker_init(void)
+void ticker_init_test()
 {
+    intFlag = 0;
+
     intf->init();
-    uint32_t previous = intf->read();
+
+    /* Wait a while - let the ticker to count. */
+    wait_ms(1);
+
+    const uint32_t ticks_start = intf->read();
+
+    intf->set_interrupt(ticks_start + TICKER_INT_VAL);
+
+    /* Re-initialise the ticker. */
+    intf->init();
+
+    const uint32_t ticks_after_reinit = intf->read();
+
+    /* Wait long enough to fire ticker interrupt (should not be fired). */
+    while (intf->read() < (ticks_start + 2 * TICKER_INT_VAL)) {
+        /* Just wait. */
+    }
+
+    TEST_ASSERT(intf->read() >= (ticks_start + 2 * TICKER_INT_VAL));
+    TEST_ASSERT(ticks_start > ticks_after_reinit);
+    TEST_ASSERT_EQUAL(0, intFlag);
+}
+
+/* Test that ticker frequency is non-zero and counter is at least 8 bits */
+void ticker_info_test(void)
+{
+    const ticker_info_t* p_ticker_info = intf->get_info();
+
+    TEST_ASSERT(p_ticker_info->frequency != 0);
+    TEST_ASSERT(p_ticker_info->bits >= 8);
+}
+
+/* Test that ticker interrupt fires only when the ticker counter increments to the value set by ticker_set_interrupt. */
+void ticker_interrupt_test(void)
+{
+    intFlag = 0;
 
     intf->init();
-    uint32_t current = intf->read();
-    TEST_ASSERT_TRUE_MESSAGE(previous <= current, "init() changed the counter");
 
-    previous = intf->read();
+    const uint32_t tick_count = intf->read();
+
+    /* Set interrupt. Interrupt should be fired when tick count is equal to:
+     * tick_count + TICKER_INT_VAL. */
+    intf->set_interrupt(tick_count + TICKER_INT_VAL);
+
+    /* Wait until ticker count reach value: tick_count + TICKER_INT_VAL - TICKER_DELTA.
+     * Interrupt should not be fired. */
+    while (intf->read() < (tick_count + TICKER_INT_VAL - TICKER_DELTA)) {
+        /* Indicate failure if interrupt has fired earlier. */
+        TEST_ASSERT_EQUAL_INT_MESSAGE(0, intFlag, "Interrupt fired too early");
+    }
+
+    /* Wait until ticker count reach value: tick_count + TICKER_INT_VAL + TICKER_DELTA.
+     * Interrupt should be fired after this time. */
+    while (intf->read() < (tick_count + TICKER_INT_VAL + TICKER_DELTA)) {
+        /* Just wait. */
+    }
```

Comment:
This test should probably be more rigorous. If the timer has not yet reached the time you scheduled the interrupt for (if the interrupt is not scheduled for the past) then the interrupt must fire. This must be true even if you are scheduled to fire 1 tick in the future.

Roughly something like this would be good:
```c
for (uint32_t delta = 1; delta < 20; delta++) {
    intFlag = 0;
    uint32_t start = us_ticker_read();
    us_ticker_set_interrupt(start + delta);
    uint32_t end = us_ticker_read();
    if (end - start < delta) {
        // wait for interrupt here
       TEST_ASSERT_EQUAL(1, intFlag);
    }
}
```

Reply:
> This must be true even if you are scheduled to fire 1 tick in the future.

I agree, but it might be hard to verify. The relation between CPU clock (time needed for execution of CPU instructions) and ticker clock frequency is various on different boards. There may be lots of combinations. So it is very possible that when `delta` from above example is a small value, then set_interrupt is called with the time in past and interrupt is not executed.
I have made the following test:
```
    static uint32_t v1 = intf->read();
    static uint32_t v2 = intf->read();

    printf("---> read1: %d \n", v1);
    printf("---> read2: %d \n", v2);

K64F:
read1: 1465162
read2: 1465167

NUCLEO-F070RB:
read1: 1673243
read2: 1673258
```

Reply:
Slightly modified this test case. I'm not sure if the following condition is now to rigorous:
```
        /* Wait until ticker count reach value: tick_count + ticker_timeout[i].
         * Interrupt should be fired after this time. */
        while (intf->read() < (tick_count + ticker_timeout[i])) {
            /* Just wait. */
        }

        TEST_ASSERT_EQUAL(1, intFlag);
```